### PR TITLE
fix: replace referer http header by dedicated class QgsHttpHeader

### PR DESCRIPTION
**Pull request by benoitdm-oslandia**
## Description

This PR provides an improvement in http header management. 

Currently only the referer http header is handled in the UI when a connection to web service is needed. This management is made by duplicating the same UI components and the same c++ behavior. Furthermore it is not present each time it may be needed (f.e., new connection to *simple* vector layer does not provide it neither raster layer nor wfs layer).

Capture d’écran du 2021-09-14 11-50-52

Also see #43951 (thanks @DelazJ).

In the case you need to define an other http header for a connection, it is not possible to set its value in the UI.

## Development 

I have created a class to handle http headers by wrapping a QMap. `QgsHttpHeaders` provides functions to :
* update request http headers
* update qgssettings
* import from qgssettings

To support this class in the UI, I have built a widget with the previous field 'referer' and with a collapsible table to set other key/value pairs.

I have replace the old http referer field in the `qgsnewarcgisrestconnection` widget by the new one. I have done the same with the `qgsvectortileconnection`.

Capture d’écran du 2021-09-14 11-50-23

If the new behavior is validated I should change all the other web service connection widget to create an uniform way to manage http headers.

An other drawback in the current way to handle the http header, is the way it is used in the `QgsDataSourceUri`. The *referer* is set as an url param but it could be an instance of QgsHttpHeaders.

### CODE REVIEW

**Comment by DelazJ** [comment]
From a GUI perspective:
- can we keep "referer" label and text edit widgets on the same line?
- Also the + and - button might better use the icons in the authentication frame
- And I think group boxes have a title case, ie HTTP Headers, More HTTP (?) Headers (though we probably can find a less repetitive label or simply remove this one?)
- (and nitpicking, harmonize the margins?)

> If the new behavior is validated I should change all the other web service connection widget to create an **uniform way** to manage http headers.

Yes please! These dialogs are quite different from each others (#43951)

**Comment by benoitdm-oslandia** [comment]
> From a GUI perspective:
> 
>     * can we keep "referer" label and text edit widgets on the same line?

agreed!

> 
>     * Also the + and - button might better use the icons in the authentication frame

Done!

> 
>     * And I think group boxes have a title case, ie HTTP Headers, More HTTP (?) Headers (though we probably can find a less repetitive label or simply remove this one?)

I replaced it by 'Advanced'

> 
>     * (and nitpicking, harmonize the margins?)
> 

I do not seen margin pbm... but I am kind a noob in UI design :). Can you explain me this margin pbm please?

**Comment by benoitdm-oslandia** [comment]
Here is the screen shot of the modifications :

Capture d’écran du 2021-09-15 08-19-26

But I still have a border around the referer group I can not remove!

File: src/core/network/qgshttpheaders.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    void setFromSettings( const QgsSettings &settings, const QString &key = QString() );
+
+    /**
+     * \param key http header key name
+     * \return http header value
+     */
+    QVariant &operator[]( const QString &key );
+
+    /**
+     * \return the list of all http header key
```
Comment (alexbruy) [inline src/core/network/qgshttpheaders.h:92]:
```suggestion
     * \return the list of all http header keys
```

File: src/gui/qgshttpheaderwidget.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    if ( ite->compare( "referer" ) != 0 )
+    {
+      addQueryPairRow( *ite, headers[ *ite ].toString() );
+    }
+    else
+    {
+      mRefererLineEdit->setText( headers[ *ite ].toString() );
+    }
+  }
+  /* mRefererLineEdit->setText( settings.value( key + "/referer" ).toString() ); */
```
Comment (alexbruy) [inline src/gui/qgshttpheaderwidget.cpp:123]: Better to remove commented code completely

**Comment by alexbruy** [comment]
Looks good to me. But I am not very familiar with this part of the QGIS codebase.

**Comment by nyalldawson** [comment]
Is this a 3.22 thing? Or 3.24?

File: python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in [hunk 3]
```diff
@@ -32,7 +32,7 @@ Utility functions for working with ArcGIS REST services.
       ImageService,
     };
 
-    static QVariantMap retrieveUserInfo( const QString &communityUrl, const QString &user, const QString &authcfg, QString &errorTitle /Out/, QString &errorText /Out/, const QMap< QString, QString > &requestHeaders = QMap< QString, QString >(), QgsFeedback *feedback = 0 );
+    static QVariantMap retrieveUserInfo( const QString &communityUrl, const QString &user, const QString &authcfg, QString &errorTitle /Out/, QString &errorText /Out/, const QgsHttpHeaders &requestHeaders = QgsHttpHeaders(), QgsFeedback *feedback = 0 );
```
Comment (nyalldawson) [inline python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in:35]: Can you confirm that sip is correctly automatically convert string maps to the new class here? Otherwise it's an API break.,.

File: src/core/network/qgshttpheaders.cpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+//
+// QgsHttpHeaders
+//
+
+const QString QgsHttpHeaders::KEY_PREFIX = "http-header/";
+
+QgsHttpHeaders::QgsHttpHeaders()
+{
+  // nope
```
Comment (nyalldawson) [inline src/core/network/qgshttpheaders.cpp:31]: ?

File: src/core/network/qgshttpheaders.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+QgsHttpHeaders::QgsHttpHeaders( const QMap<QString, QVariant> &hdrs )
+{
+  mHeaders = hdrs;
+  mHeaders.detach(); // clone like
+}
```
Comment (nyalldawson) [inline src/core/network/qgshttpheaders.cpp:38]:
```suggestion
QgsHttpHeaders::QgsHttpHeaders( const QMap<QString, QVariant> &headers )
    : mHeaders( headers )
{
  mHeaders.detach(); // clone like
}
```

File: src/core/network/qgshttpheaders.cpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+QgsHttpHeaders::QgsHttpHeaders( const QgsSettings &settings, const QString &key )
+{
+  setFromSettings( settings, key );
+}
+
+QgsHttpHeaders::~QgsHttpHeaders()
+{
+  // nope
```
Comment (nyalldawson) [inline src/core/network/qgshttpheaders.cpp:47]:
```suggestion
```

File: src/core/network/qgshttpheaders.h [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsHttpHeaders
+{
+  public:
+
+    /**
+     * used in settings
+     */
+    static const QString KEY_PREFIX;
```
Comment (nyalldawson) [inline src/core/network/qgshttpheaders.h:42]: hide from sip

File: src/core/network/qgshttpheaders.h [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * \brief constructor from map
```
Comment (nyalldawson) [inline src/core/network/qgshttpheaders.h:45]:
```suggestion
     * \brief Constructor from map
```
etc

File: src/core/network/qgshttpheaders.h [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    QgsHttpHeaders( const QMap<QString, QVariant> &headers );
+
+    /**
+     * \brief default constructor
+     */
+    QgsHttpHeaders();
+
+    /**
+     * \brief constructor from \a QgsSettings
```
Comment (nyalldawson) [inline src/core/network/qgshttpheaders.h:56]:
```suggestion
     * \brief Constructor from QgsSettings \a settings object
```

File: src/core/network/qgshttpheaders.h [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+     * \param key
+     */
+    QgsHttpHeaders( const QgsSettings &settings, const QString &key = QString() );
+
+    virtual ~QgsHttpHeaders();
+
+    /**
+     * \brief update the \a request by adding all the http headers
+     * \param request
+     * \return true if the update succeed
```
Comment (nyalldawson) [inline src/core/network/qgshttpheaders.h:67]:
```suggestion
     * \brief Updates a \a request by adding all the HTTP headers
     * \return TRUE if the update succeed
```

Docs should be sentence case, using "Updates"/"Sets"/"Returns" etc instead of "Update"/"Return" etc.

Use "TRUE" instead of "true" (and "FALSE", "NULLPTR") so that the Python documents automatically convert these to True/False/None

File: src/core/vectortile/qgsvectortilelayerrenderer.cpp [hunk 11]
```diff
@@ -163,8 +163,8 @@ bool QgsVectorTileLayerRenderer::render()
       ctx.labelingEngine()->removeProvider( mLabelProvider );
       mLabelProvider = nullptr; // provider is deleted by the engine
     }
-
-    mRequiredLayers.unite( mLabelProvider->requiredLayers( ctx, mTileZoom ) );
+    else
```
Comment (nyalldawson) [inline src/core/vectortile/qgsvectortilelayerrenderer.cpp:166]: Was this an accident? Or unrelated fix?

File: src/gui/qgshttpheaderwidget.cpp [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgshttpheaderwidget.h"
+#include "ui_qgshttpheaderwidget.h"
+#include "qgsapplication.h"
+
+
+QgsHttpHeaderWidget::QgsHttpHeaderWidget( QWidget *parent ) :
+  QWidget( parent )
```
Comment (nyalldawson) [inline src/gui/qgshttpheaderwidget.cpp:27]:
```suggestion
QgsHttpHeaderWidget::QgsHttpHeaderWidget( QWidget *parent )
   : QWidget( parent )
```

File: src/gui/qgshttpheaderwidget.cpp [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+  btnRemoveQueryPair->setIcon( QgsApplication::getThemeIcon( QStringLiteral( "/symbologyRemove.svg" ) ) );
+  grpbxAdvanced->setCollapsed( true );
+
+  setupConnections();
+}
+
+QgsHttpHeaderWidget::~QgsHttpHeaderWidget()
+{
+  // nope
+}
```
Comment (nyalldawson) [inline src/gui/qgshttpheaderwidget.cpp:40]:
```suggestion
QgsHttpHeaderWidget::~QgsHttpHeaderWidget() = default;
```

File: src/gui/qgshttpheaderwidget.cpp [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+void QgsHttpHeaderWidget::setupConnections()
+{
+  // Action and interaction connections
```
Comment (nyalldawson) [inline src/gui/qgshttpheaderwidget.cpp:44]: Just move these directly to the constructor -- no need for a separate function

File: src/gui/qgshttpheaderwidget.cpp [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+  connect( btnAddQueryPair, &QToolButton::clicked, this, &QgsHttpHeaderWidget::addQueryPair );
+  connect( btnRemoveQueryPair, &QToolButton::clicked, this, &QgsHttpHeaderWidget::removeQueryPair );
+}
+
+void QgsHttpHeaderWidget::addQueryPairRow( const QString &key, const QString &val )
+{
+  const int rowCnt = tblwdgQueryPairs->rowCount();
+  tblwdgQueryPairs->insertRow( rowCnt );
+
+  const Qt::ItemFlags itmFlags = Qt::ItemIsEnabled | Qt::ItemIsSelectable
```
Comment (nyalldawson) [inline src/gui/qgshttpheaderwidget.cpp:54]:
```suggestion
  const Qt::ItemFlags itemFlags = Qt::ItemIsEnabled | Qt::ItemIsSelectable
```

File: src/gui/qgshttpheaderwidget.cpp [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+                                 | Qt::ItemIsEditable | Qt::ItemIsDropEnabled;
+
+  QTableWidgetItem *keyItm = new QTableWidgetItem( key );
```
Comment (nyalldawson) [inline src/gui/qgshttpheaderwidget.cpp:57]:
```suggestion
  QTableWidgetItem *keyItem = new QTableWidgetItem( key );
```

None of us will get RSI from the extra keystrokes, and it's helping out non-English native speakers :laughing:

File: src/gui/qgshttpheaderwidget.cpp [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+  keyItm->setFlags( itmFlags );
+  tblwdgQueryPairs->setItem( rowCnt, 0, keyItm );
+
+  QTableWidgetItem *valItm = new QTableWidgetItem( val );
```
Comment (nyalldawson) [inline src/gui/qgshttpheaderwidget.cpp:61]:
```suggestion
  QTableWidgetItem *valueItem = new QTableWidgetItem( val );
```

File: src/providers/arcgisrest/qgsamsprovider.cpp [hunk 18]
```diff
@@ -84,7 +84,7 @@ void QgsAmsLegendFetcher::start()
     const QString referer = dataSource.param( QStringLiteral( "referer" ) );
     QgsStringMap headers;
     if ( !referer.isEmpty() )
-      headers[ QStringLiteral( "Referer" )] = referer;
+      headers[ QStringLiteral( "referer" )] = referer;
```
Comment (nyalldawson) [inline src/providers/arcgisrest/qgsamsprovider.cpp:87]: what's the rationale behind this change?

**Comment by nyalldawson** [comment]
Can you remove the accidentally committed gpkg files and squash too please?

File: src/ui/qgshttpheaderwidget.ui [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+   <item>
+    <widget class="QGroupBox" name="mHttpGroupBox">
+     <property name="sizePolicy">
+      <sizepolicy hsizetype="Preferred" vsizetype="Minimum">
+       <horstretch>0</horstretch>
+       <verstretch>0</verstretch>
+      </sizepolicy>
+     </property>
+     <property name="title">
+      <string>HTTP headers</string>
```
Comment (nyalldawson) [inline src/ui/qgshttpheaderwidget.ui:41]:
```suggestion
      <string>HTTP Headers</string>
```

File: src/core/network/qgshttpheaders.h (hunk 1 above)
Comment (benoitdm-oslandia) [reply src/core/network/qgshttpheaders.h]: done

File: src/gui/qgshttpheaderwidget.cpp (hunk 2 above)
Comment (benoitdm-oslandia) [reply src/gui/qgshttpheaderwidget.cpp]: done

File: python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in (hunk 3 above)
Comment (benoitdm-oslandia) [reply python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in:35]: I  will check this part.

File: src/core/network/qgshttpheaders.cpp (hunk 4 above)
Comment (benoitdm-oslandia) [reply src/core/network/qgshttpheaders.cpp:31]: comment when the block is intentionally empty: no operation

File: src/core/network/qgshttpheaders.cpp (hunk 6 above)
Comment (benoitdm-oslandia) [reply src/core/network/qgshttpheaders.cpp:47]: Same as above

File: src/core/network/qgshttpheaders.h (hunk 7 above)
Comment (benoitdm-oslandia) [reply src/core/network/qgshttpheaders.h:42]: ok

File: src/core/network/qgshttpheaders.h (hunk 10 above)
Comment (benoitdm-oslandia) [reply src/core/network/qgshttpheaders.h:67]: ok thanks

File: src/gui/qgshttpheaderwidget.cpp (hunk 14 above)
Comment (benoitdm-oslandia) [reply src/gui/qgshttpheaderwidget.cpp:44]: ok

File: src/core/vectortile/qgsvectortilelayerrenderer.cpp (hunk 11 above)
Comment (benoitdm-oslandia) [reply src/core/vectortile/qgsvectortilelayerrenderer.cpp:166]: No, there is nullptr acces when you passed into the above if: `mLabelProvider = nullptr` then the pointer is dereferenced at line 167.

File: src/gui/qgshttpheaderwidget.cpp (hunk 16 above)
Comment (benoitdm-oslandia) [reply src/gui/qgshttpheaderwidget.cpp:57]: Argg.. I did not get it :sweat_smile:... what do you mean by RSI?

File: src/gui/qgshttpheaderwidget.cpp (hunk 15 above)
Comment (benoitdm-oslandia) [reply src/gui/qgshttpheaderwidget.cpp]: I have duplicated this part of code from the oauth2 widget :)

File: src/providers/arcgisrest/qgsamsprovider.cpp (hunk 18 above)
Comment (benoitdm-oslandia) [reply src/providers/arcgisrest/qgsamsprovider.cpp:87]: Everywhere else, "referer" is used without the uppercase. It is more a homogenization than a bug correction.

**Comment by benoitdm-oslandia** [comment]
> Is this a 3.22 thing? Or 3.24?

good question as the feature freeze as started. As it does not matter to me, do you prefer I change to 3.24 version tags?

**Comment by benoitdm-oslandia** [comment]
> Can you remove the accidentally committed gpkg files and squash too please?

Some of them got through my checkings :(

File: python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in (hunk 3 above)
Comment (benoitdm-oslandia) [reply python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in:35]: @nyalldawson this does not work. I have to encapsulate the map in the class 
```python
        headers = QgsHttpHeaders( {'referer' : 'http://google.com'} )
        res = QgsArcGisPortalUtils.retrieveUserInfo('http://' + endpoint, 'some_user', '', headers )
```

At the beginning this new class inherited QMap but SIP is not able to handle it!
Is there a solution to have SIP working with QMap inherited class?

**Comment by benoitdm-oslandia** [comment]
> @benoitdm-oslandia code looks good to me. However, as mentioned above, I am not very familiar with this part of the codebase, so it would be good to have someone else feedback as well.

Sorry! I forgot :)

File: python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in (hunk 3 above)
Comment (nyalldawson) [reply python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in:35]: > Is there a solution to have SIP working with QMap inherited class?

I honestly have no idea. I think it'll be simplest to just leave the method with the existing signature, and add a new method which uses the QgsHttpHeaders class

File: src/core/vectortile/qgsvectortilelayerrenderer.cpp (hunk 11 above)
Comment (nyalldawson) [reply src/core/vectortile/qgsvectortilelayerrenderer.cpp:166]: Can this be moved to a separate PR please? then we can backport if necessary

File: src/gui/qgshttpheaderwidget.cpp (hunk 16 above)
Comment (nyalldawson) [reply src/gui/qgshttpheaderwidget.cpp]: "repetitive strain injury" :laughing:

Comment (benoitdm-oslandia) [reply src/gui/qgshttpheaderwidget.cpp]: okkkk! Indeed! And I have already made the changes!

File: python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in (hunk 3 above)
Comment (benoitdm-oslandia) [reply python/core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip.in:35]: Ok... so bad inheritance is a bit buggy with SIP :disappointed: 

 I add wrap functions and I test in the python test.

File: src/core/providers/arcgis/qgsarcgisportalutils.h [hunk 20]
```diff
@@ -46,6 +47,8 @@ class CORE_EXPORT QgsArcGisPortalUtils
       ImageService, //!< ArcGIS Server image service
     };
 
+#ifndef SIP_RUN
```
Comment (nyalldawson) [inline src/core/providers/arcgis/qgsarcgisportalutils.h:50]: These could still be exposed couldn't they? Sip should handle it ok and call the correct method depending on whether a dict or the new class is passed for the argument

Comment (benoitdm-oslandia) [reply src/core/providers/arcgis/qgsarcgisportalutils.h:50]: No :sad: the sip/python can not be compiled do to the fact the functions have the same name but not the same definition. I may put different names but it looks uglier than hidding the new HttpHeader class.

Comment (nyalldawson) [reply src/core/providers/arcgis/qgsarcgisportalutils.h:50]: Hmm -- that's odd. We have plenty of other functions with the same number and different signatures, and sip is happy with those. What's the actual error you get?

Comment (benoitdm-oslandia) [reply src/core/providers/arcgis/qgsarcgisportalutils.h:50]: @nyalldawson 

Here are the errors I obtain if I remove the `#ifdef SIP_RUN`:
```
sip: core/auto_generated/providers/arcgis/qgsarcgisportalutils.sip:120: ::QgsArcGisPortalUtils::retrieveUserInfo() has overloaded functions with the same Python signature
```

Is that a bug or should I put back the `#ifdef SIP_RUN`?

Regards.

**Comment by nyalldawson** [comment]
Superseded by https://github.com/qgis/QGIS/pull/46320
